deps/0007-testing-strategy.md

```diff
@@ -0,0 +1,525 @@
+
+#### Structure
+
+Unit tests should live adjacent to the code that they test in a `tests/` directory and should be written in the language of the unit of code that they are testing. Our codebase has standardized on tools based on the language used:
+
+| Language | Unit Test Framework|
+|----------|--------------------|
+| Rust     | [cargo test](https://doc.rust-lang.org/cargo/commands/cargo-test.html) |
+| Python   | [pytest](https://docs.pytest.org/en/stable/) |
+
+Rust tests should follow the naming convention that the unit under test should appear in the test name along with any additional grouping that segments a large portion of the test, e.g. the if a test evaluated model loading specific to vSGLang in the LLM module the test name should could be `llm_model_load_sglang`. This allows selection of sglang-specific tests using `cargo test sglang`.
+
+Python-based tests use a similar structure using [PyTest Marks](https://docs.pytest.org/en/stable/how-to/mark.html). Use of the mark `@pytest.mark.unit` is **required** on any unit test.
+
+#### Performance Expectations
+
+Each test suite of a functional code unit should take fewer than 15 seconds to run, excluding time required to build. In the event that the setup or execution exceeds this threshold it is an indication that either the functional unit is too large, the dependencies are not mocked appropriately, or your code is inefficient. In any of these cases an evaluation of the design is warranted.
+
+To meet this performance criteria tests should be thread safe and written with the expectation of being executed in parallel.
+
+### Integration Tests
+
+Integration tests validate the functionality of a framework module with respect to interactions with external services or between modules. This covers both testing interactions between the Dynamo framework components and a framework component and a third-party dependency, such as NATS. An example of an integration test would be validating runtime framework module initialization which requires [NATS](https://nats.io/) and [etcd](https://etcd.io/) for functionality.
+
+The primary differentiation of integration tests from [unit tests](#unit-tests) is the use of real components and the absence of mocking of response values. Simplistic models are acceptable, e.g. toy LLM models, but they should always operate within fully-functional components of the Dynamo. The primary difference with [end-to-end](#end-to-end-tests) is integration tests are driven through APIs while end-to-end tests describe end-user flows using the the command line such as `dynamo serve`.
+
+#### Structure
+
+Integration tests are defined in the [pytest](https://docs.pytest.org/en/stable/) framework and exist in within the top-level`tests` directory in subdirectories corresponding to the framework element tested. Use of the [pytest mark](https://docs.pytest.org/en/stable/how-to/mark.html#mark) `@pytest.mark.integration` is required.
```

Comment:
@pvijayakrish - I think we want this to be next to the code and not in top level